<?xml version="1.0"?>
<robot name="multi_cams_with_table" xmlns:xacro="http://ros.org/wiki/xacro">
  <!-- Include the camera system Xacro -->
  <xacro:include filename="$(find realsense2_description)/urdf/_d435.urdf.xacro"/>

  <xacro:sensor_d435 name="camera1" topics_ns="camera1" parent="world" publish_pointcloud="true">
    <origin xyz="-2.5 -2.5 4" rpy="0 0.7864 0.7864"/>
  </xacro:sensor_d435>

  <xacro:sensor_d435 name="camera2" topics_ns="camera2" parent="world" publish_pointcloud="true">
    <origin xyz="2.5 -2.5 4" rpy="0 0.7864 2.3592"/>
  </xacro:sensor_d435>

  <xacro:sensor_d435 name="camera3" topics_ns="camera3" parent="world" publish_pointcloud="true">
    <origin xyz="-2.5 2.5 4" rpy="0 0.7864 -0.7864"/>
  </xacro:sensor_d435>

  <xacro:sensor_d435 name="camera4" topics_ns="camera4" parent="world" publish_pointcloud="true">
    <origin xyz="2.5 2.5 4" rpy="0 0.7864 3.932"/>
  </xacro:sensor_d435>

  <!-- Define the table link -->
<link name="table">
  <visual>
    <geometry>
      <box size="1.0 1.0 1.0"/>
    </geometry>
    <material name="gray_color">
      <color rgba="1 1 1 1.0"/> <!-- RGBA value for gray: R=0.5, G=0.5, B=0.5, Alpha=1.0 -->
    </material>
  </visual>
  <collision>
    <geometry>
      <box size="1.0 1.0 1.0"/>
    </geometry>
  </collision>
</link>

 <!-- Define the obstacle links -->
<link name="ob1">
  <visual>
    <geometry>
      <box size="0.2 0.15 0.3"/>
    </geometry>
    <material name="red_color">
      <color rgba="0.6 0.25 0.25 1.0"/> <!-- RGBA value for gray: R=0.5, G=0.5, B=0.5, Alpha=1.0 -->
    </material>
  </visual>
  <collision>
    <geometry>
      <box size="0.2 0.15 0.3"/>
    </geometry>
  </collision>
</link>

<link name="ob2">
  <visual>
    <geometry>
      <cylinder length="0.2" radius="0.1"/>
    </geometry>
    <material name="red_color">
      <color rgba="0.6 0.25 0.25 1.0"/> <!-- RGBA value for gray: R=0.5, G=0.5, B=0.5, Alpha=1.0 -->
    </material>
  </visual>
  <collision>
    <geometry>
      <cylinder length="0.2" radius="0.1"/>
    </geometry>
  </collision>
</link>

  <link name="world"/>
  
  <!-- Define the joints -->
  <joint name="world_to_table" type="fixed">
    <parent link="world"/>
    <child link="table"/>
    <origin xyz="0 0 0.5" rpy="0 0 0"/>
  </joint>

  <joint name="table_to_robot_base_link" type="fixed">
    <parent link="table"/>
    <child link="base_link"/>
    <origin xyz="0 0 0.5" rpy="0 0 0"/>
  </joint>

  <joint name="table_to_ob1" type="fixed">
    <parent link="table"/>
    <child link="ob1"/>
    <origin xyz="-0.3 -0.3 0.5" rpy="0 0 0"/>
  </joint>

  <joint name="table_to_ob2" type="fixed">
    <parent link="table"/>
    <child link="ob2"/>
    <origin xyz="0.2 -0.25 0.5" rpy="0 0 0"/>
  </joint>

</robot>